design-documents/remote-invoker.md

```diff
@@ -0,0 +1,631 @@
+    
+    public function getUserContext(): UserContextInterface;
+}
+```
+##### Request generator
+Create request based on the remote call data
+```
+interface RequestGeneratorInterface {
+    public function generate(
+        InvokeRequestInterface $invokeRequest
+    ): RequestInterface
+}
+```
+##### Request signer
+Sign request before sending
+```
+interface RequestSignerInterface {
+    public function sign(RequestInterface $request, InvokeRequestInterface $invokeRequest): RequestInterface;
+}
+```
+##### Transport
+Send request to a remote source.
+Returns chainable and waitable promise.
+```
+interface Transport {
+    public function send(RequestInterface $request): Promise;
+}
+```
+##### Remote call request validator
+Validates whether request's signature is correct
+```
+interface RequestValidatorInterface {
+    /**
+     * @throws ValidationException
+     */
+    public function validate(RequestInterface $request): void;
+}
+```
+##### Remote call request reader
+Read incoming requests and extract remote call request information
+```
+interface RequestReaderInterface {
+    public function read(RequestInterface $request): InvokeRequestInterface;
+}
+```
+##### Remote call info storage
+Store remote call results for retries mechanism
+```
+interface InvokeResponseRepositoryInterface {
+    public function createStarted(string $callId, int $timeout): void;
+    
+    public function saveResponse(string $callId, InvokeResponseInterface $response): void;
+    
+    /**
+     * @throws NoSuchEntityException When there were no such call registered.
+     * @return InvokeResponseInterface|null Null if call is not yet finished.
+     */
+    public function get(string $callId): ?InvokeResponseInterface;
+}
+```
+##### Remote call processor
+Perform remote call
+```
+interface LocalInvokerInterface {
+    public function invoke(InvokeRequestInterface $request): InvokeResponseInterface;
+}
+```
+##### Response generator
+Transform invoke response to a regular response
+```
+interface ResponseGeneratorInterface {
+    public function generate(string $callId, InvokeResponseInterface $response): ResponseInterface;
+}
+```
+
+
+### Networking gateway
+
+_Remote code invoker_ is an abstraction, it's not enforcing how requests and responses are actually sent and parsed.
+
+#### RPC-style gateway
+
+HTTP protocol will be used. A controller will be added to process a single URL like _magento-base-url_/invoker.
+Requests' body will contain requested class name, method name, arguments and it all encoded in json.
+Requests' headers will contain information such as authenticated user ID, call ID, issued timestamp, signature.
+Response body will contain serialized service call result or a serialized thrown exception encoded in json.
+
+#### Why RESTful web API should not be used as the network gateway for the invoker
+* Not all service contracts are exposed as web API and nor should they be (for security and workflow reasons)
```

Comment:
We will have BFF that would expose public API. API of the services can be exposed and will be protected by private network.